### ISSUE TITLE
StatefulSet - can't rollback from a broken state

### ISSUE THREAD

**Issue by MrTrustor** [kubernetes/kubernetes#67250]
/kind bug

**What happened**:

I updated a StatefulSet with a non-existent Docker image. As expected, a pod of the statefulset is destroyed and can't be recreated (ErrImagePull). However, when I change back the StatefulSet with an existing image, the StatefulSet doesn't try to remove the broken pod to replace it by a good one. It keeps trying to pull the non-existing image.
You have to delete the broken pod manually to unblock the situation.

[Related Stackoverflow question](https://stackoverflow.com/questions/48894414/kubernetes-statefulset-pod-startup-error-recovery)

**What you expected to happen**:

When rolling back the bad config, I expected the StatefulSet to remove the broken pod and replace it by a good one.

**How to reproduce it (as minimally and precisely as possible)**:

1. Deploy the following StatefulSet:
```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  selector:
    matchLabels:
      app: nginx # has to match .spec.template.metadata.labels
  serviceName: "nginx"
  replicas: 3 # by default is 1
  template:
    metadata:
      labels:
        app: nginx # has to match .spec.selector.matchLabels
    spec:
      terminationGracePeriodSeconds: 10
      containers:
      - name: nginx
        image: k8s.gcr.io/nginx-slim:0.8
        ports:
        - containerPort: 80
          name: web
        volumeMounts:
        - name: www
          mountPath: /usr/share/nginx/html
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: "standard"
      resources:
        requests:
          storage: 10Gi
```

2. Once the 3 pods are running, update the StatefulSet spec and change the image to `k8s.gcr.io/nginx-slim:foobar`
3. Observe the new pod failing to pull the image.
4. Roll back the change.
5. Observe the broken pod not being deleted.

**Anything else we need to know?**:

* I observed this behaviour both on 1.8 and 1.10.
* This seems related to the discussion in #18568

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.7", GitCommit:"dd5e1a2978fd0b97d9b78e1564398aeea7e7fe92", GitTreeState:"clean", BuildDate:"2018-04-19T00:05:56Z", GoVersion:"go1.9
.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.5-gke.3", GitCommit:"6265b9797fc8680c8395abeab12c1e3bad14069a", GitTreeState:"clean", BuildDate:"2018-07-19T23:02:51Z", GoVersi
on:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: Google Kubernetes Engine
- OS (e.g. from /etc/os-release): COS

cc @joe-boyce

**Comment by joe-boyce** [comment]
Anybody have any ideas on this one?

**Comment by mattmb** [comment]
+1, I just discovered this and had assumed that it would work more like the Deployment controller.

In https://github.com/yelp/paasta we are programmatically creating/updating Deployments and StatefulSets. For that to make sense I really want them to *always* attempt to converge to the definition.

**Comment by mattmb** [comment]
Heh, well it was worth a go I suppose...

**Comment by huzhengchuan** [comment]
+1  meet the same problem
I think need to rollback success. but now it blocked

**Comment by dave-tock** [comment]
+1. This is a pretty big landmine in using StatefulSet, if you ever make any mistake you're stuck with just destroying your StatefulSet and starting over. IOW, if you ever make a mistake with StatefulSet, you need to cause an outage to recover :(

**Comment by enisoc** [comment]
I apologize for letting this bug slip through the cracks, and for implying in [my earlier comment](https://github.com/kubernetes/kubernetes/issues/67250#issuecomment-414754525) that the current behavior is expected and acceptable. My intention was to report that I can confirm the bug exists and that it appears to be a design flaw rather than a regression.

As @MrTrustor noted in the original post, the workaround for this particular incarnation of the problem is to revert to a good Pod template, and then delete the stuck Pod. As a first step, I'll send a PR to the [StatefulSet docs](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) to warn about this pitfall and describe the workaround.

The larger, unsolved problem is what to do when StatefulSet encounters a situation in which it can't be confident that any action it takes is likely to make things better instead of worse. StatefulSet was designed to provide automation for a "most-sensitive common denominator" of stateful apps, so it very often chooses to stop and do nothing if anything isn't going perfectly according to plan.

A [thread on Twitter](https://twitter.com/dave_universetf/status/1104973833088520193) discussing this bug resulted in a suggestion by @danderson for a feature to help mitigate these situations: provide an "escape hatch" to tell StatefulSet to proceed anyway, even if it detects something wrong. I think that's a good starting point for making progress on the general problem. Perhaps it could take the form of a field in the [RollingUpdate strategy](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#update-strategies) that applies all the time whenever it's set, or alternatively a one-time "acknowledge and proceed" signal that would tell StatefulSet to attempt to force-fix just one Pod but then revert to normal behavior after that.

For this specific incarnation of the general problem (a Pod that has never made it to Running state), we might be able to do something more automatic. Perhaps we can argue that it should always be safe to delete and replace such a Pod, as long as none of the containers (even init containers) ever ran. We'd also need to agree on whether this automatic fix-up can be enabled by default without being considered a breaking change, or whether it needs to be gated by a new field in StatefulSet that's off by default (until `apps/v2`).

I'll start investigating some of these options and report back any findings. If anyone else has ideas or input, please leave them here.

**Comment by andpol** [comment]
To add a bit more information to this discussion, it's worth noting that the issue described in the original post is not present if you specify `podManagementPolicy: Parallel` in the StatefulSet spec. From what I can tell, this is because the value of `monotonic` is true in the code quoted [above](https://github.com/kubernetes/kubernetes/issues/67250#issuecomment-414754525). I guess this is somewhat implied by the policy name (rather calling it `ParallelReady` or something), and the StatefulSet docs do hint at this, but it was still a surprise to me when performing a rolling update.

**Comment by enisoc** [comment]
@andpol Thanks, that's a good point. I made sure to mention it in my PR documenting this known issue: https://github.com/kubernetes/website/pull/13190

**Comment by janetkuo** [comment]
This works as intended, and is documented here:

https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#updating-statefulsets
> The StatefulSet controller terminates each Pod, and **waits for it to transition to Running and Ready prior to updating the next Pod.** Note that, even though the StatefulSet controller will not proceed to update the next Pod until its ordinal successor is Running and Ready, it will restore any Pod that fails during the update to its current version. Pods that have already received the update will be restored to the updated version, and Pods that have not yet received the update will be restored to the previous version. In this way, the controller attempts to continue to keep the application healthy and the update consistent in the presence of intermittent failures.

**Comment by danderson** [comment]
I think we're talking about different things. The case I hit seemed to be a straightforward logic bug: if you create a StatefulSet with a broken template (in my case, resource requests such that all pods remain pending for ever), the sts controller will ignore any changes to the StatefulSet from there on. I applied a corrected template that could definitely schedule, deleted all pods... And the StatefulSet recreated them using the broken template, where they went pending again. I had to delete the entire StatefulSet and recreate it to get my template changes to apply.

That's what I meant when I spoke of overrides to @enisoc . While it's accepted that manual intervention will be needed sometimes (either by a human or a software operator), what I encountered was a situation where the StatefulSet ignored my interventions, and kept deploying broken pods. That was my complaint, not "StatefulSet cannot guess what is safe or not for my arbitrary stateful workload", that part definitely makes sense.

**Comment by enisoc** [comment]
@danderson Ah that's different. Can you give some more details to help me understand and reproduce the form of the bug you hit?

I tried the following steps with k8s v1.12.1:

1. I created a StatefulSet apps/v1 with a memory request of `100Gi`.
2. StatefulSet created pod-0 which was stuck Pending due to resource request.
3. I edited the StatefulSet to remove the memory request from the Pod template.
4. StatefulSet did nothing. This is the form of the bug we've discussed above.
5. I deleted pod-0. StatefulSet recreated pod-0 without the memory request (using updated template), and then proceeded to create the rest of the Pods. This is the workaround discussed above.

Since you said in your case "all pods" (implying more than one) were created but Pending, I suspected you may be using `podManagementPolicy: Parallel` so I tried the above steps with that setting:

1. I created a StatefulSet apps/v1 with a memory request of `100Gi` and `podManagementPolicy: Parallel`.
2. StatefulSet created pod-0, pod-1, pod-2 but all were stuck Pending due to resource requests.
3. I edited the StatefulSet to remove the memory request from the Pod template.
4. StatefulSet deleted the Pending pod-2 and recreated it with the new Pod template, which became Running and Ready. Then it did the same for pod-1, followed by pod-0. This is consistent with the [comment above](https://github.com/kubernetes/kubernetes/issues/67250#issuecomment-472988799) that this bug should not manifest when the podManagementPolicy is Parallel.

Do you get different results? Or are there steps that I did differently?

**Comment by antonlisovenko** [comment]
This is a real blocker for programmatic usage of Statefulsets (from the Operator for example). The real use case: the statefulset is created with some memory/cpu limits which cannot be fulfiled. So say 2 pods are running and the 3rd is staying Pending as it cannot be scheduled to the node as there no available resources. Trying to fix and change the statefulset specification to have smaller limits doesn't help - the statefulset specification is updated but all the pods stay unchanged forever as the 3rd pod is Pending. The only way is to delete the pod manually which contradicts the nature of operators totally

**Comment by enisoc** [comment]
After some initial investigation, I believe the mitigation described above should be feasible:

> For this specific incarnation of the general problem (a Pod that has never made it to Running state), we might be able to do something more automatic. Perhaps we can argue that it should always be safe to delete and replace such a Pod, as long as none of the containers (even init containers) ever ran. We'd also need to agree on whether this automatic fix-up can be enabled by default without being considered a breaking change, or whether it needs to be gated by a new field in StatefulSet that's off by default (until apps/v2).

The specific incarnation we're talking about is when all of the following are true:

1. There's a Pod stuck Pending because it was created from a bad StatefulSet revision.
2. Deleting that stuck Pod (i.e. the workaround discussed above) would result in the Pod being replaced at a different revision (meaning we have reason to expect a different result; we won't hot-loop).

In this situation, I think we can argue that it's safe for StatefulSet to delete the Pending Pod for you, as long as we can ensure the Pod has not started running before we get around to deleting it. The argument would be, if the Pod never ran, then the application should not be affected one way or another if we delete it. We could potentially use the new [ResourceVersion precondition on Delete](https://github.com/kubernetes/kubernetes/pull/74040) to give a high confidence level that the Pod never started running.

There is still a very slight chance that a container started running and had some effect on the application already but the kubelet has not updated the Pod status yet. However, I would argue that the chance of that is small enough that we should take the risk in order to prevent StatefulSet from getting stuck in this common situation.

I probably won't have time to work on the code for this any time soon since I'm about to change jobs. However, I'm willing to commit to being a reviewer if anyone is available to work on this.

**Comment by TopherGopher** [comment]
@enisoc I'm going to try taking a whack at this

**Comment by draveness** [comment]
This issue is similar to #78007

**Comment by dims** [comment]
@alejandrox1 @krmayankk Do you want to deal with this in the tech debt work group? PTAL

**Comment by tyrken** [comment]
What can we do to get this going again?  It sounds like @janetkuo, @enisoc & probably others feel it's unsafe to delete a Pending pod if it's started running at all - at least until we can introduce a new explicit safe-to-proceed flag like:

```yaml
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      # outdatedPendingPods: waitForManualDeletion  # default as now
      outdatedPendingPods: safeToReplaceViaDeletion  # new proposed aggressive functionality 
      partition: <n>
```

If this can't happen until an apps/v2, then can we come up with a new test a bit like #78709 by @TopherGopher but stricter than IsPending, outdated & the right side of a partition?

I've observed that sounds like a pod never having been scheduled (e.g. when stuck due to insufficient resources/anti-affinities), or all a scheduled pod's containers/init-containers in state `waiting` with `restartCount == 0` (e.g. invalid image).  The `restartCount` restriction is necessary to avoid crash-looping main/init containers which go back to `waiting` while in back-off, but therefore have run once so need manual action (pod-deletion).

Note I haven't managed to replicate the @danderson situation, apart from something similar when the `partition` number was set to a high value, locking the pods to the old revision even after individual deletion.

**Comment by skyfirst93** [comment]
It is this fixed ?

**Comment by piyushkumar13** [comment]
I also ran into same issue... is there any timeline to fix this issue or any alternative available ?

**Comment by wu0407** [comment]
+1

**Comment by clong-** [comment]
Still seeing this issue on v1.20.5.

**Comment by sidh** [comment]
Aaaand we just stepped on this landmine too. :(

**Comment by ialidzhikov** [comment]
/remove-lifecycle stale

@janetkuo , can you add `lifecycle/frozen` to this issue?

**Comment by sftim** [comment]
From https://github.com/kubernetes/kubernetes/issues/67250#issuecomment-472650229:
>  alternatively a one-time "acknowledge and proceed" signal that would tell StatefulSet to attempt to force-fix just one Pod but then revert to normal behavior after that.

A new subresource perhaps? Maybe a `RollingUpdateOverride`? If so, I assume that'd need its own KEP.

**Comment by watkinsmike** [comment]
This continues to bite us. How are people managing this in the meantime? The manual intervention of deleting the stuck pod is not very ideal.

**Comment by zerkms** [comment]
> How are people managing this in the meantime?
manually :-(

**Comment by skorobogatydmitry** [comment]
+1

**Comment by feilengcui008** [comment]
+1, any update or plan for this problem?

**Comment by ashwiniramesha** [comment]
+1  this is causing unintended bookkeeping of the statefulset names. and maintaining this information across product releases is becoming an overhead.

**Comment by joinvimals** [comment]
+1 
Hi kubernetes team,
any update on this ?  fixing the issue will help a lot

**Comment by April-Q** [comment]
+1
some issue in 1.19

**Comment by euqen** [comment]
hey guys! 

also getting issues with updating statefulsets in broken state :(

**Comment by yifan-gu-anchorage** [comment]
Hitting the same issue in k8s version 1.21.6

**Comment by kerthcet** [comment]
/assign 
I'll take a look as if this is highly needed by the community.

**Comment by kerthcet** [comment]
FYI: KEP initialized here https://github.com/kubernetes/enhancements/pull/3562, glad to hear any advices.

**Comment by teqwve** [comment]
It might be worth noting that `podManagementPolicy: Parallel` workaround does not seem to work when the feature gate `MaxUnavailableStatefulSet` is enabled (at least on v1.30).

**Comment by soltysh** [comment]
This is currently working as intended, see [documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#forced-rollback). I suggest to track work in https://github.com/kubernetes/enhancements/issues/3541 for eventual progress.

/close

**Comment by helayoty** [comment]
Update for everyone following this issue: after several design iterations, [KEP-3541] (https://github.com/kubernetes/enhancements/3541) has been implemented and ships as **alpha in Kubernetes v1.37**, behind the `StatefulSetRecreateStrategy` feature gate.

Implementation: [kubernetes/kubernetes#137187](https://github.com/kubernetes/kubernetes/pull/137187)

## The solution: a new `Recreate` update strategy

StatefulSets now support a third update strategy, mirroring the Deployment `Recreate` strategy:

```yaml
spec:
  updateStrategy:
    type: Recreate
```

When the pod template changes, the controller deletes **all** pods, waits for them to fully terminate, and then creates pods from the new revision according to `podManagementPolicy`. Old and new revisions never run at the same time.

This resolves the scenario in this issue: with `Recreate`, a bad revision cannot wedge the workload. Push a broken image, roll back, so the controller deletes the stuck pods and brings everything up on the corrected revision, with no manual `kubectl delete pod`. Since switching between strategies is allowed, you can also unblock a StatefulSet that is already stuck mid-`RollingUpdate` by switching it to `Recreate`. The feature gate must be enabled on both kube-apiserver and kube-controller-manager.

The trade-off is explicit: every update takes all replicas down. The default `RollingUpdate` behavior is unchanged.

## Why not auto-heal `RollingUpdate` directly?

The original proposal (`EnforcedRollingUpdate`) did exactly what this issue asks for: the controller would automatically delete a not-Ready pod stuck on an outdated revision. It didn't survive API review, for two reasons:

1. **Safety for stateful workloads.** Automatically deleting a stuck pod is only provably safe if none of its containers (including init containers) ever ran. That excludes the common CrashLoopBackOff and mid-startup-crash cases, and even the "never ran" check races kubelet status reporting. For workloads that own data, a controller deleting a pod that may have run is a potential data-loss bug, not a convenience. The current `RollingUpdate` behavior of restoring failed pods at their existing revision is intentional and exists to protect exactly these workloads.
2. **Backward compatibility.** Changing *when* `RollingUpdate` deletes pods is a semantics change to a GA API that existing workloads depend on. Hiding the new behavior behind an optional timeout was considered and rejected as too surprising.

A separate opt-in strategy resolves both: you consent to deletion semantics explicitly when you set `type: Recreate`, and every existing StatefulSet is untouched by construction. A bounded auto-recovery option for `RollingUpdate` (e.g. a timeout) was deliberately deferred as a possible future extension.

This is alpha. Please test it and report feedback on [kubernetes/enhancements#3541](https://github.com/kubernetes/enhancements/3541) since real-world usage is what moves it to beta.
